### PR TITLE
#26 Implement getCurrentSingleSignOnAccount() livecycle aware

### PR DESCRIPTION
**Pull request by stefan-niedermann**
- Add `CurrentSingleSignOnAccountLiveData` class which registers a `OnSharedPreferencesChanged` listener in a lifecycle aware way (It basically does only emit new events when the `Activity` / `Fragment` is not `paused` or `stopped`
- Add `SingleAccountHelper#getCurrentSingleSignOnAccount$` which returns a `CurrentSingleSignOnAccountLiveData` instance. The trailing `$` is a convention borrowed from `rxjs`[<sup>[1]</sup>](https://angular.io/guide/rx-library#naming-conventions-for-observables) to signalize that this is something observable
- `@Deprecate`ed `SingleAccountHelper#(un)registerSharedPreferenceChangeListener` methods, which are not lifecycle aware and potential memory leaks when used wrong
- Add sample usage to sample app

Example usage from 3rd party app:

```java
SingleAccountHelper.getCurrentSingleSignOnAccount$(this).observe(this, ssoAccount -> {
    Log.i(TAG, "New SingleSignOnAccount set: " + ssoAccount.name);
});
```

Signed-off-by: Stefan Niedermann <info@niedermann.it>

### CODE REVIEW

**Comment by David-Development** [comment]
This looks great! I haven't personally worked with Live data before but definetely a great addition and perfect use case! I think we should probably mention it in the docs, eh?  Not sure if devs will find it if it isn't documented in the Readme..?

**Comment by stefan-niedermann** [comment]
The PR is nust to demonstrate the general idea - of course we can add some docs and discuss the API before merging this. 🙂

**Comment by stefan-niedermann** [comment]
I have added minimal documentation in the `README.md` and in the sample app :slightly_smiling_face:

**Comment by stefan-niedermann** [comment]
Would love to get a review @AndyScherzinger @David-Development :slightly_smiling_face:

File: lib/src/main/java/com/nextcloud/android/sso/helper/SingleSignOnAccountLiveData.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    private SingleSignOnAccount getValueFromPreferences() throws NextcloudFilesAppAccountNotFoundException, NoCurrentAccountSelectedException {
+        return getCurrentSingleSignOnAccount(context);
+    }
+
+    @Override
+    protected void onInactive() {
+        sharedPrefs.unregisterOnSharedPreferenceChangeListener(preferenceChangeListener);
+        super.onInactive();
+    }
+}
```
Comment (AndyScherzinger) [inline lib/src/main/java/com/nextcloud/android/sso/helper/SingleSignOnAccountLiveData.java:74]: missing new line at the end of the file

**Comment by stefan-niedermann** [comment]
More changes since the last review:

- In case of an error, the `LiveData<SingleSignOnAccount>` no emits <code>null</code> instead of nothing
- Added `@WorkerThread` annotations to synchronous methods
- Aligned method names (`setCurrentSingleSignOnAccount` → `applyCurrentSingleSignOnAccount` / `commitCurrentSingleSignOnAccount`), backwards compatible
- Added some more documentation

**Comment by stefan-niedermann** [comment]
Actually, reading from `SharedPreferences` *is* synchronously. It ain't "heavy", but it is. This is the reason why the [`DataStore`](https://developer.android.com/topic/libraries/architecture/datastore) framework (on top of `SharedPreferences`) promotes the async operations and explicitly warns about using the synchronous APIs.

**Pull request merged**